
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ohimjosh/CSC-44800/blob/main/AS04/Assignment%204%20-%20Sample.ipynb)




In [ ]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn import datasets
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm 
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [ ]:
# Load the dataset (load remotely, not locally)
iris = datasets.load_iris()
data = pd.DataFrame(iris.data, columns= iris.feature_names)
# Output the first 15 rows of the data
data.head(15)
# Display a summary of the table information (number of datapoints, etc.)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                 5.1               3.5                1.4               0.2
1                 4.9               3.0                1.4               0.2
2                 4.7               3.2                1.3               0.2
3                 4.6               3.1                1.5               0.2
4                 5.0               3.6                1.4               0.2
5                 5.4               3.9                1.7               0.4
6                 4.6               3.4                1.4               0.3
7                 5.0               3.4                1.5               0.2
8                 4.4               2.9                1.4               0.2
9                 4.9               3.1                1.5               0.1
10                5.4               3.7                1.5               0.2
11                4.8               3.4                1.6               0.2
12                4.8               3.0                1.4               0.1
13                4.3               3.0                1.1               0.1
14                5.8               4.0                1.2               0.2

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


In [ ]:
data.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The Iris flower data set contains 150 entries with 50 samples each of 3 flower species. The 3 flower species are the labels in the data which are Iris Setosa, Iris Virginica and Iris versicolor. Also included are the four features sepal length, sepal width, petal length and petal width, all measured in centimeters. The labels are mapped to the actual classes with 0 as Iris Setosa, 1 as Iris Virginica and 2 as Iris versicolor.


# Part 2: Split the dataset into train and test

In [ ]:
# Take the dataset and split it into our features (X) and label (y)
X = iris.data   #features
y = iris.target #label
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size= 0.9, test_size=0.1)

In [ ]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [ ]:
X_train

array([[4.4, 3.2, 1.3, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5.1, 3.5, 1.4, 0.2],
       [6.3, 3.3, 6. , 2.5],
       [5.5, 2.5, 4. , 1.3],
       [4.9, 3.6, 1.4, 0.1],
       [7.7, 2.8, 6.7, 2. ],
       [6.5, 3. , 5.5, 1.8],
       [5.5, 2.4, 3.8, 1.1],
       [6.9, 3.1, 5.4, 2.1],
       [6.3, 2.5, 5. , 1.9],
       [7.3, 2.9, 6.3, 1.8],
       [6.4, 2.8, 5.6, 2.1],
       [4.8, 3. , 1.4, 0.1],
       [6.7, 3.1, 5.6, 2.4],
       [7. , 3.2, 4.7, 1.4],
       [5.1, 3.7, 1.5, 0.4],
       [5.6, 2.7, 4.2, 1.3],
       [5.1, 3.4, 1.5, 0.2],
       [6.1, 3. , 4.6, 1.4],
       [5.7, 3.8, 1.7, 0.3],
       [5.2, 3.4, 1.4, 0.2],
       [5. , 3.5, 1.6, 0.6],
       [7.9, 3.8, 6.4, 2. ],
       [5.6, 2.5, 3.9, 1.1],
       [6. , 3.4, 4.5, 1.6],
       [6.7, 2.5, 5.8, 1.8],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.4, 1.5, 0.4],
       [5. , 3.6, 1.4, 0.2],
       [5.9, 3. , 5.1, 1.8],
       [6.4, 3.2, 5.3, 2.3],
       [4.6, 3.6, 1. , 0.2],
       [4.6, 3.4, 1.4, 0.3],
       [5.8, 2

In [ ]:
X_test

array([[7.4, 2.8, 6.1, 1.9],
       [6.3, 2.3, 4.4, 1.3],
       [6.1, 2.8, 4. , 1.3],
       [6.1, 2.8, 4.7, 1.2],
       [5.8, 2.7, 5.1, 1.9],
       [5. , 3.3, 1.4, 0.2],
       [5.6, 2.9, 3.6, 1.3],
       [5.7, 4.4, 1.5, 0.4],
       [6. , 2.2, 5. , 1.5],
       [7.6, 3. , 6.6, 2.1],
       [5.2, 4.1, 1.5, 0.1],
       [5.1, 3.8, 1.9, 0.4],
       [7.2, 3.6, 6.1, 2.5],
       [6.3, 3.3, 4.7, 1.6],
       [7.7, 3. , 6.1, 2.3]])

In [ ]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [ ]:
y_train

array([0, 0, 0, 2, 1, 0, 2, 2, 1, 2, 2, 2, 2, 0, 2, 1, 0, 1, 0, 1, 0, 0,
       0, 2, 1, 1, 2, 0, 0, 0, 2, 2, 0, 0, 1, 2, 1, 1, 2, 2, 1, 1, 0, 1,
       1, 0, 1, 2, 0, 2, 2, 1, 1, 2, 0, 1, 2, 0, 0, 0, 1, 2, 0, 0, 0, 2,
       0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 2, 1, 2, 2, 2, 1, 2, 0, 0, 0, 0,
       0, 2, 1, 0, 1, 2, 0, 2, 1, 1, 0, 1, 1, 2, 0, 2, 2, 0, 2, 0, 1, 1,
       2, 1, 2, 1, 1, 0, 0, 0, 2, 2, 2, 0, 1, 1, 1, 2, 2, 2, 2, 1, 0, 1,
       0, 1, 2])

In [ ]:
y_test


array([2, 1, 1, 1, 2, 0, 1, 0, 2, 2, 0, 0, 2, 1, 2])

# Part 3: Logistic Regression

In [ ]:

# i. Use sklearn to train a LogisticRegression model on the training set
logreg = LogisticRegression().fit(X_train, y_train)
logreg.predict(X_test[0].reshape(1, -1))

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for Logistic regression model, what does the score measure?

# iv. Extract the coefficents and intercepts for the boundary line(s)

array([2])

In [ ]:
logreg_pred = logreg.predict(X_test[0].reshape(1, -1))
logreg_pred

array([2])

In [ ]:
logreg.score(X_test, y_test)

0.9333333333333333

In [ ]:
logreg.coef_

array([[-0.41464563,  0.89397135, -2.43772401, -1.03168631],
       [ 0.50045064, -0.46118722, -0.15990522, -0.95137971],
       [-0.08580501, -0.43278413,  2.59762923,  1.98306603]])

In [ ]:
logreg.intercept_

array([  9.66159328,   2.58146512, -12.24305839])

### Score for Logistic regression model

The score returned from the test is 0.9333 which is a 93.33% accuracy. This is a high score but the goal is to reach 1.0 which is a 100% accuracy.

# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
clfsvm = svm.SVC(probability=True).fit(X_train, y_train)
clfsvm.fit(X_train, y_train)
clfsvm.predict(X_test)
# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for the SVM, what does the score measure?

array([2, 1, 1, 1, 2, 0, 1, 0, 2, 2, 0, 0, 2, 1, 2])

In [ ]:
clfsvm_pred = clfsvm.predict(X_test[0].reshape(1, -1))
clfsvm_pred


array([2])

In [ ]:
clfsvm.score(X_test, y_test)

1.0

### Score for Support Vector Machine model

The score returned from the test is 1.0 which is a 100% accuracy. This is a perfect score which we are aiming for. 

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
clfneural = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1)
clfneural.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
clfneural_pred = clfneural.predict(X_test[0].reshape(1, -1))
clfneural_pred

# iii. Report on the score for the Neural Network, what does the score measure?

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

array([0])

In [ ]:
clfneural.score(X_test, y_test)

0.26666666666666666

In [ ]:
clfneural2 = MLPClassifier(solver='lbfgs', hidden_layer_sizes=(3, 3))
clfneural2.fit(X_train, y_train)
clfneural2_pred = clfneural2.predict(X_test[0].reshape(1, -1))
clfneural2_pred

array([2])

In [ ]:
clfneural2.score(X_test, y_test)

1.0

### Score for Support Vector Machine model

The score returned from the test is 0.2666 which is a 26.66% accuracy. This is a low score that is far from our target. I changed the parameters for the second neural network and was able to get a score of 1.0, only with solver='lbfgs'and hidden_layer_sizes=(3, 3).

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
neigh = KNeighborsClassifier()
neigh.fit(X_train, y_train)
neigh_pred = neigh.predict(X_test)

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for kNN, what does the score measure?

In [ ]:
neigh_pred = neigh.predict(X_test[0].reshape(1, -1))
neigh_pred

array([2])

In [ ]:
neigh.score(X_test, y_test)

0.9333333333333333

### Score for K-Nearest Neighbors model

The score returned from the test is 1.0 which is a 100% accuracy. This is a really good model since we received a perfect score. 

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The results in the notebook show the different machine learning models on the iris dataset. It allows us to see how these models interact with the data given and find out the best model. The best models performed on the data set is K-nearest Neighbors and neural network with the right parameters. One thing that I was confused about it every time I would run the cells a new score would be given for another model. This may be because of my local machine or there must be a bug somewhere. Written in the markdown is the score that I saw at the time.